packages/grafana-ui/src/types/datasource.ts

```diff
@@ -173,8 +173,53 @@ export type LegacyResponseData = TimeSeries | TableData | any;
 
 export type DataQueryResponseData = SeriesData | LegacyResponseData;
 
+export type DataStreamObserver = (event: DataStreamEvent) => void;
+
+export interface DataStreamEvent {
+  /**
+   * Consistent key across events
+   */
+  key: string;
+
+  /**
+   * when Done or Error no more events will be processed
+   */
+  state: LoadingState;
+
+  /**
+   * The initial request for this stream
+   */
+  request: DataQueryRequest;
+
+  /**
+   * Series data
+   */
+  series: SeriesData[];
+
+  /**
+   * Error in stream (but may still be running)
+   */
+  error?: DataQueryError;
+
+  /**
+   * Optionally return only the rows that changed
+   */
+  delta?: SeriesData[];
+
+  /**
+   * Callback function to stop the data stream
+   */
+  shutdown: () => void;
+}
+
 export interface DataQueryResponse {
   data: DataQueryResponseData[];
+
+  /**
+   * When streaming, this must return the list of expected events
+   * other events will be ignored and shutdown
+   */
+  streams?: DataStreamEvent[];
```

Comment:
the naming here feels very confusing, the obserer takes events, but the request method should return expected events? Is this more like a DataStreamState rather than an event?

Reply:
This is the part I am least happy with -- yes DataStreamState is better.

The key issues we need to solve are:
* keeping consistent order as stream events arrive (otherwise the series colors change!)
* surviving a refresh event intelligently
* closing streams reliably -- when done or query changes

In the approach above, we:
1. return a list of `DataStreamState` from the ds.query() -- close open streams that don't match open keys.
2. as events arrive, it checks that they match a `key` from the latest response, otherwise close them

I would much rather keep the `ds.query()` result unchanged, and somehow manage order and stale streams with just the callbacks and query method.  I think its possible, but have not figured it out yet :)